test/e2e/kubelet.go

```diff
@@ -87,25 +90,81 @@ func waitTillNPodsRunningOnNodes(c *client.Client, nodeNames sets.String, podNam
 	})
 }
 
+func updateNodes(c *client.Client, nodeNames sets.String, labels map[string]string, add bool) {
+	const maxRetries = 5
+	for nodeName := range nodeNames {
+		var node *api.Node
+		var err error
+		for i := 0; i < maxRetries; i++ {
+			node, err = c.Nodes().Get(nodeName)
+			if err != nil {
+				framework.Logf("Error getting node %s: %v", nodeName, err)
+				continue
+			}
+			for k, v := range labels {
+				if add {
+					node.ObjectMeta.Labels[k] = v
+				} else {
+					delete(node.ObjectMeta.Labels, k)
+				}
+			}
+			_, err := c.Nodes().Update(node)
+			if err != nil {
+				framework.Logf("Error updating node %s: %v", nodeName, err)
+			} else {
+				break
+			}
+		}
+		Expect(err).NotTo(HaveOccurred())
+	}
+}
+
 var _ = framework.KubeDescribe("kubelet", func() {
+	var c *client.Client
 	var numNodes int
 	var nodeNames sets.String
+	var nodeSelector map[string]string
 	f := framework.NewDefaultFramework("kubelet")
 	var resourceMonitor *framework.ResourceMonitor
 
 	BeforeEach(func() {
 		nodes := framework.GetReadySchedulableNodesOrDie(f.Client)
 		numNodes = len(nodes.Items)
 		nodeNames = sets.NewString()
-		for _, node := range nodes.Items {
-			nodeNames.Insert(node.Name)
+		// If there are a lot of nodes, we don't want to use all of them
+		// (if there are 1000 nodes in the cluster, starting 10 pods/node
+		// will take ~10 minutes today). And there is also deletion phase.
+		//
+		// Instead, we choose at most 10 nodes and will constraint pods
+		// that we are creating to be scheduled only on that nodes.
+		if numNodes > maxNodesCount {
```

Comment:
Actually, after applying other comments, the code is much cleaner. And applying those labels always will not reduce number of ifs, because:
- we need this one to check how many nodes we have
- we need one for starting resourceMonitor, because we don't want to monitor e.g. 2000 nodes.

And all other disappeared after applying other comments :)